<a href="https://colab.research.google.com/github/sdasmaha/Ml_Engineers_445/blob/master/Heart_Notebook_Exercise1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

 Notebook Excercise #1 - SRINJOY DASMAHAPATRA

 IDE => Google Colab


#Importing Libraries

In [167]:
import numpy as np
import pandas as pd

#Mounting the Google Drive


In [168]:
from google.colab import drive
drive.mount("/content/gdrive")

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


#Loading the Dataset

In [169]:
heart_df = pd.read_csv('/content/gdrive/My Drive/Colab Notebooks/BAJWA-ML/heart_failure_dataset.csv')

# *1.1*

**A**

In [155]:
print(heart_df.shape) 

(299, 13)


In [156]:
print(heart_df.axes)

[RangeIndex(start=0, stop=299, step=1), Index(['age', 'anaemia', 'creatinine_phosphokinase', 'diabetes',
       'ejection_fraction', 'high_blood_pressure', 'platelets',
       'serum_creatinine', 'serum_sodium', 'sex', 'smoking', 'time',
       'DEATH_EVENT'],
      dtype='object')]


In [157]:
print(heart_df.dtypes)

age                         float64
anaemia                       int64
creatinine_phosphokinase      int64
diabetes                      int64
ejection_fraction             int64
high_blood_pressure           int64
platelets                   float64
serum_creatinine            float64
serum_sodium                  int64
sex                           int64
smoking                       int64
time                          int64
DEATH_EVENT                   int64
dtype: object


**B**

In [174]:
heart_df.head(10)

age  anaemia  creatinine_phosphokinase  ...  smoking  time  DEATH_EVENT
0   75.0        0                       582  ...        0     4            1
1   55.0        0                      7861  ...        0     6            1
2   65.0        0                       146  ...        1     7            1
3   50.0        1                       111  ...        0     7            1
4   65.0        1                       160  ...        0     8            1
5   90.0        1                        47  ...        1     8            1
6   75.0        1                       246  ...        0    10            1
7   60.0        1                       315  ...        1    10            1
8   65.0        0                       157  ...        0    10            1
9   80.0        1                       123  ...        1    10            1
10  75.0        1                        81  ...        1    10            1
11  62.0        0                       231  ...        1    10            1
12  45.0        1                       981  ...        0    11            1
13  50.0        1                       168  ...        0    11            1
14  49.0        1                        80  ...        0    12            0
15  82.0        1                       379  ...        0    13            1
16  87.0        1                       149  ...        0    14            1
17  45.0        0                       582  ...        0    14            1
18  70.0        1                       125  ...        0    15            1
19  48.0        1                       582  ...        0    15            1

[20 rows x 13 columns]

**C**
    Each row of heart_df dataframe is termed called a data sample associated with an instance.

**D**    Based on the knowledge of the dataset and its stated usage, we are dealing with supervised problem. 

Since in a supervised learning model, the algorithm learns on a labeled dataset, providing an answer key that the algorithm can use to evaluate its accuaracy on training data. 

On the contrary the unsupervised model works on unlabeled data that the algorithm tries to make sense of by extracting features and patterns on its own.


**E** The independent variables in this case are  labels- 

1. age
2. anaemia
3. creatinine_phosphokinase
4. diabetes
5. ejection_fraction
6. high_blood_pressure
7. platelets
8.	serum_creatinine
8. serum_sodium
9. sex
10. smoking
11. time

This is so because these variable/labels embedded in the algorithm 
increase or decrease the probability/likhehood of the event(*DEATH_EVENT*).


**F** The dependent variable in this case is the label *DEATH_EVENT* since it is the conclusive/final event or variable that predicts the likelihood of death.

**G** Here the categorical variables are -


1. anaemia
2. diabetes 
3. high_blood_pressure
4. sex
5. smoking
6. DEATH_EVENT


This is so because a categorical variable is a variable that can take on one of a limited, and usually fixed, number of possible values, assigning each individual or other unit of observation to a particular group or nominal category on the basis of some qualitative property.



**H** The categorical variables follow *L Encoding* where each label is mapped to a binary number.

In [ ]:
count1 = heart_df['DEATH_EVENT'].value_counts() 
print(count1)

0    203
1     96
Name: DEATH_EVENT, dtype: int64


**I** 


1.   Samples corresponding to Deceased Patients : 96
2.   Samples corresponding to Remaining Patients : 203



In [ ]:
count2 = heart_df['sex'].value_counts() 
print(count2)

1    194
0    105
Name: sex, dtype: int64


**J** 

1.   Samples corresponding to Women Patients : 105
2.   Samples corresponding to Male Patients : 194


In [ ]:
count3 = heart_df['smoking'].value_counts() 
print(count3)

0    203
1     96
Name: smoking, dtype: int64


**K** 

1.   Samples corresponding to Smokers  : 96
2.   Samples corresponding to Non-Smokers : 203


#*1.2*

Side Notes

1.   Positive Correlation = A positive relationship exists i.e. when the indepenedent variable increases in value, so does the dependent variable

2.  Negative Correlation = A negative relationship exists i.e. when the independent variable increases in value, the dependent variable falls in value



In [ ]:
heart_df.corr()

age   anaemia  ...      time  DEATH_EVENT
age                       1.000000  0.088006  ... -0.224068     0.253729
anaemia                   0.088006  1.000000  ... -0.141414     0.066270
creatinine_phosphokinase -0.081584 -0.190741  ... -0.009346     0.062728
diabetes                 -0.101012 -0.012729  ...  0.033726    -0.001943
ejection_fraction         0.060098  0.031557  ...  0.041729    -0.268603
high_blood_pressure       0.093289  0.038182  ... -0.196439     0.079351
platelets                -0.052354 -0.043786  ...  0.010514    -0.049139
serum_creatinine          0.159187  0.052174  ... -0.149315     0.294278
serum_sodium             -0.045966  0.041882  ...  0.087640    -0.195204
sex                       0.065430 -0.094769  ... -0.015608    -0.004316
smoking                   0.018668 -0.107290  ... -0.022839    -0.012623
time                     -0.224068 -0.141414  ...  1.000000    -0.526964
DEATH_EVENT               0.253729  0.066270  ... -0.526964     1.000000

[13 rows x 13 columns]

**A** The two variables that are the most positively correlated with the *DEATH-EVENT* are -

1.   serum_creatinine (0.294278)
2.   age (0.253729)





**B** The two variables that are the most negatively correlated with the *DEATH-EVENT* are -

1.   time (-0.526964	)
2.   ejection_fraction (-0.268603	)






**C** The answer to this question can be understood by general intuition and research conducted in the past. With increasing age, DEATH_EVENT is likely to increase.  

**D** Past Research has shown with decrease in ejection fraction(Percentage of blood leaving the heart at each contraction) there is an increase in the likelihood of Death (DEATH_EVENT).

#*1.3*


1.Checking for missing values.


In [ ]:
heart_df.isnull().sum()

age                         0
anaemia                     0
creatinine_phosphokinase    0
diabetes                    0
ejection_fraction           0
high_blood_pressure         0
platelets                   0
serum_creatinine            0
serum_sodium                0
sex                         0
smoking                     0
time                        0
DEATH_EVENT                 0
dtype: int64

We can see here that we are not dealing with missing values.

2. Checking for Inconsistent values. For this we have to assess each columns and categorize them first into their data types and then search for inconsistency.

We divide columns into categorical and non-categorical values.


Numerical = NUM ; Categorical = C
1. age - NUM
2. anaemia - C
3. creatinine_phosphokinase - NUM
4. diabetes - C
5. ejection_fraction - NUM
6. high_blood_pressure - C
7. platelets - NUM
8.	serum_creatinine -NUM
8. serum_sodium - NUM
9. sex - C
10. smoking - C
11. time - NUM
13. DEATH_EVENT - C

For Categorical values

In [14]:
heart_df['anaemia'].value_counts()

0    170
1    129
Name: anaemia, dtype: int64

In [6]:
heart_df['diabetes'].value_counts()

0    174
1    125
Name: diabetes, dtype: int64

In [7]:
heart_df['high_blood_pressure'].value_counts()

0    194
1    105
Name: high_blood_pressure, dtype: int64

In [8]:
heart_df['sex'].value_counts()

1    194
0    105
Name: sex, dtype: int64

In [21]:
heart_df['smoking'].value_counts()

0    203
1     96
Name: smoking, dtype: int64

In [9]:
heart_df['DEATH_EVENT'].value_counts()

0    203
1     96
Name: DEATH_EVENT, dtype: int64

Conclusion => All the categorical values are consistent.

For NUMERICAL values


In [ ]:
pd.cut(x=heart_df['age'], bins = [-100,0,100]).value_counts()

(0, 100]     299
(-100, 0]      0
Name: age, dtype: int64

In [22]:
pd.cut(x=heart_df['creatinine_phosphokinase'], bins=[-1000,0, 20, 40, 60, 80, 100,1000]).value_counts()

(100, 1000]    203
(60, 80]        24
(80, 100]       17
(40, 60]        17
(20, 40]         2
(0, 20]          0
(-1000, 0]       0
Name: creatinine_phosphokinase, dtype: int64

In [23]:
pd.cut(x=heart_df['ejection_fraction'], bins=[-1000,0, 20, 40, 60, 80, 100,1000]).value_counts()

(20, 40]       196
(40, 60]        75
(0, 20]         23
(60, 80]         5
(100, 1000]      0
(80, 100]        0
(-1000, 0]       0
Name: ejection_fraction, dtype: int64

In [30]:
pd.cut(x=heart_df['platelets'], bins=[-100000,0, 50000, 100000, 150000, 200000,250000 ,300000,400000,1000000]).value_counts()

(250000, 300000]     84
(200000, 250000]     72
(300000, 400000]     60
(150000, 200000]     35
(400000, 1000000]    20
(100000, 150000]     20
(50000, 100000]       6
(0, 50000]            2
(-100000, 0]          0
Name: platelets, dtype: int64

In [31]:
pd.cut(x=heart_df['serum_creatinine'], bins=[-10,0, 1, 3, 5, 7,9,11, 13,20 ]).value_counts()

(1, 3]      154
(0, 1]      131
(3, 5]        9
(5, 7]        3
(9, 11]       1
(7, 9]        1
(13, 20]      0
(11, 13]      0
(-10, 0]      0
Name: serum_creatinine, dtype: int64

In [40]:
pd.cut(x=heart_df['serum_sodium'], bins=[-1000,0,100,120,122,124,126,128,130,132,134,136,138,140,200,1000]).value_counts()

(136, 138]     61
(138, 140]     57
(134, 136]     56
(140, 200]     42
(132, 134]     42
(130, 132]     19
(128, 130]     11
(126, 128]      5
(124, 126]      2
(100, 120]      2
(122, 124]      1
(120, 122]      1
(200, 1000]     0
(0, 100]        0
(-1000, 0]      0
Name: serum_sodium, dtype: int64

In [41]:
pd.cut(x=heart_df['time'], bins=[-1000,0, 20, 40, 60, 80, 100,1000]).value_counts()

(100, 1000]    172
(80, 100]       35
(60, 80]        29
(20, 40]        25
(0, 20]         23
(40, 60]        15
(-1000, 0]       0
Name: time, dtype: int64

Hence we can confirm there are no inconsistent values. 



#*1.4*


Non Categorical Independent Variable. = Numerical Variables (NUM)

We have to convert the NUM variables in the dataset have zero mean and unit variance. This process of conversion is sometimes referred to as standardization.


 ***standard score = (observed value-mean of the sample) / standard deviation of the sample***


In [94]:
heart_df_NC = heart_df[['age','creatinine_phosphokinase','ejection_fraction','platelets','serum_creatinine','serum_sodium','time']]

In [95]:
avg = heart_df_NC.mean()

In [96]:
stdv = heart_df_NC.std()

In [97]:
standardized = (heart_df_NC - avg) / stdv

In [111]:
standardized.head(20)

age  creatinine_phosphokinase  ...  serum_sodium      time
0   1.190949                  0.000165  ...     -1.501519 -1.626775
1  -0.490457                  7.502063  ...     -0.141739 -1.601007
2   0.350246                 -0.449186  ...     -1.728149 -1.588122
3  -0.910808                 -0.485257  ...      0.084892 -1.588122
4   0.350246                 -0.434757  ...     -4.674340 -1.575238
5   2.452003                 -0.551217  ...     -1.048259 -1.575238
6   1.190949                 -0.346124  ...      0.084892 -1.549470
7  -0.070106                 -0.275011  ...     -1.274889 -1.549470
8   0.350246                 -0.437849  ...      0.311522 -1.549470
9   1.611300                 -0.472890  ...     -0.821629 -1.549470
10  1.190949                 -0.516176  ...     -1.274889 -1.549470
11  0.098035                 -0.361583  ...      0.764782 -1.549470
12 -1.331160                  0.411384  ...      0.084892 -1.536586
13 -0.910808                 -0.426512  ...      0.084892 -1.536586
14 -0.994879                 -0.517207  ...      0.311522 -1.523701
15  1.779441                 -0.209051  ...     -0.141739 -1.510817
16  2.199792                 -0.446094  ...      0.764782 -1.497933
17 -1.331160                  0.000165  ...     -2.181409 -1.497933
18  0.770597                 -0.470829  ...      0.764782 -1.485049
19 -1.078949                  0.000165  ...     -3.541189 -1.485049

[20 rows x 7 columns]

In [115]:
standardized.mean()

age                         6.735601e-16
creatinine_phosphokinase   -2.135044e-17
ejection_fraction           4.752794e-17
platelets                  -1.114679e-15
serum_creatinine            4.026879e-16
serum_sodium               -8.420429e-16
time                       -6.052386e-16
dtype: float64

In [148]:
standardized.var()

age                         1.0
creatinine_phosphokinase    1.0
ejection_fraction           1.0
platelets                   1.0
serum_creatinine            1.0
serum_sodium                1.0
time                        1.0
dtype: float64

We can clearly see the mean is ~ 0 and the variance is 1 for all the non categorical columns.

#*1.5*


In [170]:
one_hot = pd.get_dummies(heart_df['DEATH_EVENT'])

In [171]:
heart_df1 = heart_df.drop('DEATH_EVENT', axis =1)

In [172]:
heart_df1 = heart_df1.join(one_hot)

In [173]:
heart_df1.head(20)

age  anaemia  creatinine_phosphokinase  diabetes  ...  smoking  time  0  1
0   75.0        0                       582         0  ...        0     4  0  1
1   55.0        0                      7861         0  ...        0     6  0  1
2   65.0        0                       146         0  ...        1     7  0  1
3   50.0        1                       111         0  ...        0     7  0  1
4   65.0        1                       160         1  ...        0     8  0  1
5   90.0        1                        47         0  ...        1     8  0  1
6   75.0        1                       246         0  ...        0    10  0  1
7   60.0        1                       315         1  ...        1    10  0  1
8   65.0        0                       157         0  ...        0    10  0  1
9   80.0        1                       123         0  ...        1    10  0  1
10  75.0        1                        81         0  ...        1    10  0  1
11  62.0        0                       231         0  ...        1    10  0  1
12  45.0        1                       981         0  ...        0    11  0  1
13  50.0        1                       168         0  ...        0    11  0  1
14  49.0        1                        80         0  ...        0    12  1  0
15  82.0        1                       379         0  ...        0    13  0  1
16  87.0        1                       149         0  ...        0    14  0  1
17  45.0        0                       582         0  ...        0    14  0  1
18  70.0        1                       125         0  ...        0    15  0  1
19  48.0        1                       582         1  ...        0    15  0  1

[20 rows x 14 columns]